charmbits/httpservice/service.go

```diff
@@ -208,191 +327,120 @@ func (svc *Service) Restart() error {
-	}
-	certPEM := args[2]
-	h, err := srv.handler.get(args[3])
-	if err != nil {
-		return nil, errgo.Newf("cannot get handler: %v", err)
-	}
-	httpListener, httpAddr, err := srv.listenHTTP(httpPort)
-	if err != nil {
-		return nil, errgo.Mask(err)
-	}
-	if httpsPort != 0 && certPEM != "" {
-		httpsListener, httpsAddr, err := srv.listenHTTPS(httpsPort, certPEM)
-		if err != nil {
-			httpListener.Close()
-			return nil, errgo.Mask(err)
-		}
-		srv.httpsListener = httpsListener
-		httpsServer := &http.Server{
-			Addr:    httpsAddr,
-			Handler: h,
-		}
-		srv.tomb.Go(func() error {
-			return httpsServer.Serve(httpsListener)
-		})
-	}
-
-	srv.httpListener = httpListener
-	httpServer := &http.Server{
-		Addr:    httpAddr,
-		Handler: h,
-	}
-	srv.tomb.Go(func() error {
-		return httpServer.Serve(httpListener)
-	})
-	srv.tomb.Go(func() error {
-		<-srv.tomb.Dying()
-		if srv.httpListener != nil {
-			srv.httpListener.Close()
-		}
-		if srv.httpsListener != nil {
-			srv.httpsListener.Close()
-		}
-		return nil
-	})
-	return srv, nil
-}
-
-// Kill implements hook.Command.Kill.
-func (srv *server) Kill() {
-	srv.tomb.Kill(nil)
-}
-
-// Wait implements hook.Command.Wait.
-func (srv *server) Wait() error {
-	return srv.tomb.Wait()
-}
-
-func (*server) listenHTTP(port int) (net.Listener, string, error) {
-	addr := ":" + strconv.Itoa(port)
-	listener, err := net.Listen("tcp", addr)
-	if err != nil {
-		return nil, "", errgo.Newf("cannot listen on %s: %v", addr, err)
-	}
-	return listener, addr, nil
-}
-
-func (*server) listenHTTPS(port int, certPEM string) (net.Listener, string, error) {
-	certPEMBytes := []byte(certPEM)
-	cert, err := tls.X509KeyPair(certPEMBytes, certPEMBytes)
-	if err != nil {
-		return nil, "", errgo.Newf("cannot parse certificate: %v", err)
-	}
-	config := &tls.Config{
-		NextProtos:   []string{"http/1.1"},
-		Certificates: []tls.Certificate{cert},
-	}
-	addr := ":" + strconv.Itoa(port)
-	listener, err := net.Listen("tcp", addr)
-	if err != nil {
-		return nil, "", errgo.Newf("cannot listen on %s: %v", addr, err)
-	}
-	return tls.NewListener(
-		tcpKeepAliveListener{listener.(*net.TCPListener)},
-		config,
-	), addr, nil
-}
-
-type handler struct {
-	argType reflect.Type
-	fv      reflect.Value
+type handlerInfo struct {
+	argType      reflect.Type
+	relationType reflect.Type
+	fv           reflect.Value
 }
 
 var (
-	httpHandlerType = reflect.TypeOf((*http.Handler)(nil)).Elem()
-	errorType       = reflect.TypeOf((*error)(nil)).Elem()
+	handlerType = reflect.TypeOf((*Handler)(nil)).Elem()
+	errorType   = reflect.TypeOf((*error)(nil)).Elem()
 )
 
-func newHandler(f interface{}) (*handler, error) {
+// newHandler makes a new handler value from the given function,
+// which should be in one of the forms expected by Service.Register.
+// It also registers any relations required by the handler.
+func (svc *Service) newHandlerInfo(f interface{}, registry *hook.Registry) (*handlerInfo, error) {
```

Comment:
Please also state which fields are accepted (or not accepted, like from todo anonymous fields).